src/main/cpp/src/hive_hash.cu

```diff
@@ -210,11 +208,219 @@ class hive_device_row_hasher {
       return this->hash_functor.template operator()<T>(col, row_index);
     }
 
+    /**
+     * @brief A structure representing an element in the stack used for processing columns.
+     *
+     * This structure is used to keep track of the current column being processed, the index of the
+     * next child column to process, and current hash value for the column.
+     */
+    struct col_stack_element {
+     private:
+      cudf::column_device_view _column;  // current column
+      hive_hash_value_t _cur_hash;       // current hash value of the column
+      int _child_idx;  // index of the child column to process next, initialized as 0
+
+     public:
+      __device__ col_stack_element() =
+        delete;  // Because the default constructor of `cudf::column_device_view` is deleted
+
+      __device__ col_stack_element(cudf::column_device_view col)
+        : _column(col), _child_idx(0), _cur_hash(HIVE_INIT_HASH)
+      {
+      }
+
+      __device__ void update_cur_hash(hive_hash_value_t child_hash)
+      {
+        _cur_hash = _cur_hash * HIVE_HASH_FACTOR + child_hash;
+      }
+
+      __device__ hive_hash_value_t get_hash() { return _cur_hash; }
+
+      __device__ int get_and_inc_child_idx() { return _child_idx++; }
+
+      __device__ int cur_child_idx() { return _child_idx; }
+
+      __device__ cudf::column_device_view get_column() { return _column; }
+    };
+
+    typedef col_stack_element* col_stack_element_ptr;
+
+    /**
+     * @brief Functor to compute the hive hash value for a nested column.
+     *
+     * This functor produces the same result as "HiveHash" in Spark for nested types.
+     * The pseudocode of Spark's HiveHash function is as follows:
+     *
+     * hive_hash_value_t hive_hash(NestedType element) {
+     *    hive_hash_value_t hash = HIVE_INIT_HASH;
+     *    for (int i = 0; i < element.num_child(); i++) {
+     *        hash = hash * HIVE_HASH_FACTOR + hive_hash(element.get_child(i));
+     *    }
+     *    return hash;
+     * }
```

Comment:
This algorithm only applies to struct and non-nested types. When we have lists, which can have many elements, what will we do with them? And in case lists of lists of....of lists, thing will be very complicated.

Reply:
Here is the [Spark Hive Hash Implementation](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L920), where the computation logic for `lists` and `structs` is similar.

Reply:
> Ref: Here is cudf code for hashing nested elements: https://github.com/rapidsai/cudf/blob/d67d017e8c3176ae214165869d5bb4b02c4207a0/cpp/include/cudf/table/experimental/row_operators.cuh#L1926-L1937

I checked this, seems it's not applicable to hive hash.
e.g.: hive hash for: [[1,2], [3,4,5]] ,
It expects:
`h1 = 31 * hash(1) + hash(2); h2 = 31 * (31 * hash(3) + hash(4)) + hash(5);  final_hash = 31 * h1 + h2`
Above code in `row_operators.cuh` has better performance, but it's logic is:
First update hash according sub list sizes: [2, 3], then combine for each item in each sub list.
For hive hash, [[1,2], [3,4,5]] and [[1,2,3], [4,5]] have diff hash values. We need to use stack to record the intermediate hash of sub list.

@ustcfy  inspired by the approach `row_operators` Nghia pointed,  it's not optimal to put all the ternimal item into stack. E.g.: put 1, 2, 3, 4, 5 to stack, then compute the hash value.
We can detect if current element is list of basic types, if it is, then use a `for` iterator to compute hash directly:
```c++
if (curr_list_element.child() is not nested type) {
  val hash = SEED;
  for (int i = 0; i < size; ++i) {
    // compute for basic type, then accumulate
    hash = 31 * hash + hash_dispatch(curr_element.child(i));
  }
}
```

Reply:
> @ustcfy inspired by the approach `row_operators` Nghia pointed, it's not optimal to put all the ternimal item into stack. E.g.: put 1, 2, 3, 4, 5 to stack, then compute the hash value. We can detect if current element is list of basic types, if it is, then use a `for` iterator to compute hash directly:
> 
> ```c++
> if (curr_list_element.child() is not nested type) {
>   val hash = SEED;
>   for (int i = 0; i < size; ++i) {
>     // compute for basic type, then accumulate
>     hash = 31 * hash + hash_dispatch(curr_list_element.slice(i));
>   }
> }
> ```
Plan to add these changes in the next release.